File: src/scenes/test_scene.cpp
#include "wforge/assets.h"
#include "wforge/scene.h"
#include <format>
#include <proxy/v4/proxy.h>

namespace wf::scene {

namespace {

constexpr int WIDTH = 256;
constexpr int HEIGHT = 192;

} // namespace

TestScene::TestScene()
	: font(AssetsManager::instance().getAsset<PixelFont>("font"))
	, _duck_texture(
		  AssetsManager::instance().getAsset<sf::Texture>("duck/texture")
	  ) {}

std::array<int, 2> TestScene::size() const {
	return {WIDTH, HEIGHT};
}

void TestScene::setup(SceneManager &mgr) {
	mgr.setWindowTitle("Native Test Scene");
}

void TestScene::handleEvent(SceneManager &mgr, sf::Event &evt) {
	if (auto kb = evt.getIf<sf::Event::KeyPressed>()) {
		if (kb->code == sf::Keyboard::Key::Escape) {
			mgr.changeScene(pro::make_proxy<SceneFacade, MainMenu>());
		}
	}
}

void TestScene::step(SceneManager &mgr) {
	_frame_count++;
	_x += _dx;
	_y += _dy;
	if (_x <= 0 || _x >= WIDTH - static_cast<int>(_duck_texture.getSize().x)) {
		_dx = -_dx;
	}
	if (_y <= 0 || _y >= HEIGHT - static_cast<int>(_duck_texture.getSize().y)) {
		_dy = -_dy;
	}
}

void TestScene::render(
	const SceneManager &mgr, sf::RenderTarget &target, int scale
) const {
	sf::RectangleShape top_bar(sf::Vector2f(WIDTH * scale, 20 * scale));
	top_bar.setPosition(sf::Vector2f(0, 0));
	top_bar.setFillColor(sf::Color(50, 50, 80));
	target.draw(top_bar);

	font.renderText(
		target, "Native Test Scene", sf::Color(200, 200, 50), 10, 3, scale
	);
	font.renderText(
		target, "Press ESC for menu", sf::Color(150, 150, 150), 10, 24, scale
	);
	font.renderText(
		target, std::format("Frame: {}", _frame_count),
		sf::Color(100, 100, 100), 10, HEIGHT - 16, scale
	);

	sf::Sprite duck_sprite(_duck_texture);
	duck_sprite.setPosition(sf::Vector2f(_x * scale, _y * scale));
	duck_sprite.setScale(sf::Vector2f(scale, scale));
	target.draw(duck_sprite);
}

} // namespace wf::scene

File: ts-src/test_scene.ts
/// <reference path="./waveforge.d.ts" />

const WIDTH = 256;
const HEIGHT = 192;
let frameCount = 0;
let x = 50, y = 50, dx = 2, dy = 1;

const duck = new waveforge.Texture("duck/texture");
const cmds = new waveforge.DrawCmdList();
const drawDuckCmd = new waveforge.DrawSpriteCmd(duck, x, y);
const bgRect = new waveforge.DrawRectCmd(0, 0, WIDTH, HEIGHT, new waveforge.Color(200, 220, 255));
const redRect = new waveforge.DrawRectCmd(10, 10, 40, 30, new waveforge.Color(255, 100, 100));
const fpsText = new waveforge.DrawTextCmd("FRAME: 0", 4, 4, 1, new waveforge.Color(0, 0, 0));

bgRect.r = 220;
bgRect.g = 240;
bgRect.b = 255;

cmds.push(bgRect);
cmds.push(redRect);
cmds.push(drawDuckCmd);
cmds.push(fpsText);

waveforge.log("Loaded " + duck.width + "x" + duck.height + " duck texture");
waveforge.log("Test scene loading...");

waveforge.setupScene({
    size() {
        return [WIDTH, HEIGHT];
    },

    setup() {
        waveforge.log("Test scene setup complete");
    },

    step() {
        frameCount++;
        x += dx;
        y += dy;
        if (x <= 0 || x >= WIDTH - duck.width) dx = -dx;
        if (y <= 0 || y >= HEIGHT - duck.height) dy = -dy;
        drawDuckCmd.x = x;
        drawDuckCmd.y = y;
        fpsText.text = "FRAME: " + frameCount;
    },

    handleEvent(event: waveforge.SceneEvent) {
		if (event.type === "mousemove") {
			return; // Too noisy
		}
        waveforge.log("Event: " + event.type + ("code" in event ? " code=" + event.code : ""));
    },

    render() {
        waveforge.commitDraw(cmds);
    },
});
